src/cargo/ops/cargo_rustc/context/mod.rs

```diff
@@ -246,11 +246,17 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     /// all the units mentioned in `units`.
     pub fn probe_target_info(&mut self) -> CargoResult<()> {
         debug!("probe_target_info");
-        self.probe_target_info_kind(Kind::Target)?;
-        if self.requested_target().is_none() {
+        let host_target_same = match self.requested_target() {
+            Some(s) if s != self.config.rustc()?.host => false,
+            _ => true,
+        };
+
+        if host_target_same {
+            self.probe_target_info_kind(Kind::Target)?;
             self.host_info = self.target_info.clone();
         } else {
             self.probe_target_info_kind(Kind::Host)?;
+            self.probe_target_info_kind(Kind::Target)?;
```

Comment:
Oh, actually this side-effectual mutating API looks horrible to me :) Do you feel like refactoring it to

```
self.host_info = self.probe_target_info_kind(Kind::Host);
self.target_info = if host_target_same { self.host_info.clone() } else { self.probe_target_info_kind(Kind::Target) }
```

?